athenz/resource_role.go

```diff
@@ -169,24 +201,76 @@ func resourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interfac
 	return nil
 }
 
+// return true iff the members is includes in the members list
+func isMembersIncludes(member *zms.RoleMember, members []*zms.RoleMember) bool {
+	for _, m := range members {
+		if string(member.MemberName) == string(m.MemberName) {
+			return true
+		}
+	}
+	return false
+}
+
+// return all role members that appears in list1 but not includes in list2
+func unifyMembers(list1, list2 []*zms.RoleMember) []*zms.RoleMember {
+	toReturn := make([]*zms.RoleMember, 0)
+	for _, m := range list1 {
+		if !isMembersIncludes(m, list2) {
+			toReturn = append(toReturn, m)
+		}
+	}
+	return toReturn
+}
+
+/*
+since we not allow configuring both members (deprecated) and member attributes at the same time, state changes CAN'T be one of the following:
+1. remove members from both attributes member and members.
+2. add members in both attributes member and members.
+*/
+func handleMembersChange(removeDeprecatedRoleMembers, addDeprecatedRoleMembers, removeRoleMembers, addRoleMembers []*zms.RoleMember) ([]*zms.RoleMember, []*zms.RoleMember) {
+	var removeMembers []*zms.RoleMember
+	if len(removeDeprecatedRoleMembers) == 0 {
+		removeMembers = unifyMembers(removeRoleMembers, addDeprecatedRoleMembers)
+	} else {
+		removeMembers = unifyMembers(removeDeprecatedRoleMembers, addRoleMembers)
+	}
+	if len(addDeprecatedRoleMembers) == 0 {
+		return removeMembers, addRoleMembers
+	}
+	return removeMembers, addDeprecatedRoleMembers
+}
```

Comment:
not sure what we are trying to do here

Reply:
I want to make sure not to delete unnecessary members when the users switch between members attribute to member attribute and vice versa